MM-14559 Rewrite existing plugin API tests to exclusively test via actual plugin

reflog opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
Move from testing using internal plugin APIs to testing via RPC. Previously embedded Go code is moved to separate files.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10438

mattermod commented:
Thanks @reflog for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

grundleborg commented:
/check-cla

grundleborg commented:
@crspeller @levb would be great if you can give some feedback on the approach taken here.

crspeller commented:
@reflog Looks good! Certainly heading in the right direction.
What might be cool is automatically traversing the tests folder and automatically running all the tests inside. (you would have to somehow specify your assertions and options inside the file) Anything to ease test development would encourage more tests to be written.

reflog commented:
Hi @crspeller , I've went ahead with your suggestion and implemented a 'bulk test' function that scans a folder of tests. Each plugin test is responsible for it's own assertions based on the data passed to it.

I've split all basic tests into separate files (I defined 'basic' as any test that doesn't require any preparations)

Let me know what you think!

crspeller commented:
@reflog Looking good! 
The only comment is it's a bit inconvenient to not have the usual assertion libraries available in the plugins. Maybe a test specific hook that passes the test context? 0/5 no idea if that would even work.

reflog commented:
@crspeller - I'm not sure that passing test context through a specific hook is a viable idea and would add complexity that I think might be better to avoid. This way, the plugins don't have any 'test specific' code and the assertions are done via regular 'if' statements.

What do you think? Is the ticket merge-able in it's current state, or you want to pursue the 'test context' idea?

crspeller commented:
@reflog  I think it should be mergeable without that suggestion. It's better than it was.

levb commented on app/plugin_api_test.go:
```diff
@@ -376,45 +359,10 @@ func TestPluginAPILoadPluginConfiguration(t *testing.T) {
-		func (p *MyPlugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
-			return nil, fmt.Sprintf("%v%v%v", p.configuration.MyStringSetting, p.configuration.MyIntSetting, p.configuration.MyBoolSetting)
-		}
+	pwd, _ := os.Getwd()
```
(nit) 1/5 maybe use `fileutils.FindDir("tests")` like in https://github.com/mattermost/mattermost-server/blob/master/api4/file_test.go#L30

reflog replied:
Done!

levb commented:
@reflog @crspeller Were you guys talking about passing the test context via an RPC, and invoking it remotely? I'm very curious if it works, maybe just pass it as a `testing.TB` interface so that there's no data serialization involved, or do you specifically need a `testing.T` for the assertions? Not sure if there's a difference in the mechanics, really.

reflog commented:
@levb - yes, @crspeller suggested maybe passing along the testing context by adding another RPC interface to the plugins, in order to support assertion libraries.
I think it's a little bit of an overkill, since the way it's implemented right now - each plugin test is written exactly like a regular plugin would, without any hacks and does testing 'of itself'.

I can start investigating the passing of the testing context if you feel that it's worth it. If not - let's merge :)
